fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,312 @@
+    /// The state tree of the current blockchain
+    state_tree: StateTree<ReadOnlyBlockstore<DB>>,
+    /// The latest block height with state params serialized to bytes
+    block_state_params_bytes: Vec<u8>,
+    /// Block state params cid derived from block state params bytes
+    block_state_params_cid: Cid,
+}
+
+impl<DB> V1Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
+        let state_tree_root = state_params.state_root;
+
+        let block_state_params_bytes = fvm_ipld_encoding::to_vec(&(state_params, block_height))?;
+        let block_state_params_cid =
+            Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&block_state_params_bytes));
+
+        Ok(Self {
+            state_tree_root,
+            state_tree,
+            block_state_params_bytes,
+            block_state_params_cid,
+        })
+    }
+
+    /// For V1 snapshot, we are putting two components into the CAR file: the state tree and latest
+    /// state params. One header is for the state tree root. The other header root cid is for the
+    /// serialized block state params.
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        Ok(vec![self.state_tree_root, self.block_state_params_cid])
+    }
+
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    async fn write_car(self, car: CarHeader, write: tokio::fs::File) -> anyhow::Result<()> {
+        let write_task = tokio::spawn(async move {
+            let mut streamer =
+                StateTreeStreamer::new(self.state_tree_root, self.state_tree.into_store());
+            let mut write = write.compat_write();
+
+            car.write_stream_async(&mut Pin::new(&mut write), &mut streamer)
+                .await
+                .unwrap();
+
+            let mut streamer = tokio_stream::iter(vec![(
+                self.block_state_params_cid,
+                self.block_state_params_bytes,
+            )]);
+            car.write_stream_async(&mut Pin::new(&mut write), &mut streamer)
+                .await
+                .unwrap();
+        });
+
+        write_task.await?;
+
+        Ok(())
+    }
+}
+
+#[pin_project::pin_project]
+struct StateTreeStreamer<BlockStore> {
+    /// The list of cids to pull from the blockstore
+    #[pin]
+    dfs: VecDeque<Cid>,
+    /// The block store
+    bs: BlockStore,
+}
+
+impl<BlockStore> StateTreeStreamer<BlockStore> {
+    pub fn new(state_root_cid: Cid, bs: BlockStore) -> Self {
+        let mut dfs = VecDeque::new();
+        dfs.push_back(state_root_cid);
+        Self { dfs, bs }
+    }
+}
+
+impl<BlockStore: Blockstore> Stream for StateTreeStreamer<BlockStore> {
```

Comment:
Out of a curiosity, has this implementation of the streamer been extracted from Forest? If this is the case I would add a note pointing to the original implementation (if possible).

Reply:
It's not strictly a reference, but the inspiration is from https://github.com/ChainSafe/forest/blob/77d6b2b128d73900b0162e3f573ff8d63e6324b3/src/ipld/util.rs#L335

Reply:
I usually include a comment to point the reader at the original, in case they too want to get the inspiration from the horse's mouth.